src/docs/stan-reference/appendices.tex

```diff
@@ -663,35 +664,55 @@ \subsection{Expressions}
 \small
 \begin{Verbatim}[fontsize=\small]
 expressions ::= expression % ','
-expression ::= real_literal
-             | variable
-             | '{' expressions '}'
-             | expression `?` expression `:` expression
+
+expression ::= expression `?` expression `:` expression
              | expression infixOp expression
              | prefixOp expression
              | expression postfixOp
-             | expression '[' indexes ']'
-             | function_literal '(' ?expressions ')'
-             | function_literal '(' expression ?('|' expression % ',') ')'
-             | integrate_ode '(' function_literal (',' expression){6} ')'
-             | integrate_ode_rk45
-               '(' function_literal (',' expression){6|9} ')'
-             | integrate_ode_bdf
-               '(' function_literal (',' expression){6|9} ')'
-             | algebra_solver
-             '('function_literal (',' expression){4|7} ')'
-             | '(' expression ')'
+             | common_expression
+
+range_expression ::= range_expression mathInfixOp range_expression
+                   | prefixOp range_expression
+                   | range_expression postfixOp
+                   | range_expression '[' indexes ']'
+                   | common_expression
+
+common_expression ::= real_literal
+                    | variable
+                    | '{' expressions '}'
+                    | function_literal '(' ?expressions ')'
+                    | function_literal '(' expression ?('|' expression % ',') ')'
+                    | integrate_ode '(' function_literal (',' expression){6} ')'
+                    | integrate_ode_rk45
+                      '(' function_literal (',' expression){6|9} ')'
+                    | integrate_ode_bdf
+                      '(' function_literal (',' expression){6|9} ')'
+                    | algebra_solver
+                    '('function_literal (',' expression){4|7} ')'
+                    | '(' expression ')'
+
+prefixOp ::= '!' | '-' | '+' | '^'
+
+postfixOp ::= '\\''
+
+infixOp ::= mathInfixOp | logicalInfixOp
+
+mathInfixOp ::= ('+' | '-' | '*' | '/' | '%' | '\\' | '.*' | './')
+
+logicalInfixOp :: ('||' | '&&' | '==' | '!=' | '<' | '>' | '<=' | '>=')
 
 index ::= ?(expression | expression ':' | ':' expression
-            | expression ':' expression)
+        | expression ':' expression)
 
 indexes ::= index % ','
 
-integer_literal ::= 0 | [1-9] [0-9]*
+integer_literal ::= [0-9]+
```

Comment:
The original one is correct.  Integer literals may not have leading zeroes, e.g., "007" is not an integer literal.

Reply:
But this parses and compiles.
```
library("rstan")
foo <- "transformed data {
  int a = 007;
} model {
  
}"
stan_model(model_code = foo)
```

Reply:
And so does all this
```
library("rstan")
foo <- "transformed data {
  real alpha = 0e7;
  real bravo = 0;
  int charlie = 0;
  int delta = 0000000;
  real echo = 00.;
  int foxtrot = 007;
} model {
  
}"
stan_model(model_code = foo)
```

So it seems that at the moment the current implementation doesn't care about the leading zeros. Unless it is treating them as reals and then casting them back into integers?

Reply:
Really?  That was purely by accident.  I meant to restrict to C++ literals.  Here's how the code gets compiled:

```
  stan::math::assign(alpha,0.0);
...
  stan::math::assign(echo,0.0);
...
  stan::math::assign(foxtrot,7);
```

So we must be using Boost somewhere to convert the funny numbers with leading zeroes to integers on the insde of the AST, then generating back out in standardized way.

What should we do about this?  I suppose it doesn't hurt anything other than aesthetically to allow all those crazy non-standard literals.

P.S.  Model blocks are optional now!

Reply:
I was also surprised that `model` blocks don't require `{}` and could be followed immediately by a single statement.

Reply:
So what blocks are actually required now? Are they all optional in the grammar, with any issues arising in the sematics?

Since the leading 0's don't cause any harm, I don't see why not. Is not having a leading `0` in C++ literals due to octal literals?